text/0000-layer-origin-metadata.md

```diff
@@ -0,0 +1,276 @@
+
+# Summary
+[summary]: #summary
+
+<!-- One paragraph explanation of the feature. -->
+
+Exposing layer origin metadata in `report.toml` will improve layer
+traceability and, in turn, build and image traceability.
+
+# Motivation
+[motivation]: #motivation
+
+- Layer generation can be influenced by the state of the file system,
+  namely by the state of the application directory, other layers, and
+  the stack's build image
+- Layers aren't traceable, because we can't know enough about the
+  state of the file system at the instant that the buildpack build
+  executable is invoked (TODO: figure out if there's a less wordy name
+  for this)
+- Images aren't traceable, because their constituent layers aren't
+  traceable
+- Layers and images should be traceable
+- Exposing layer origin metadata will enable us to know crucial facts
+  about the state of the file system at the time that the buildpack
+  build executable is invoked
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Terminology
+
+- **Layer origin metadata** is the missing piece of the traceability
+  puzzle; it includes the ID of the restored version of the layer (if
+  applicable), modification date, source version info, and a list of
+  the IDs of the layers that were present at layer generation time.
+- **Layer ID** is the layer's unique ID, whether on disk or on the
+  registry.
+- **Traceability** concerns the problem of identifying the inputs to
+  an image or layer (compare with **reproducibility**, which concerns
+  the problem of producing identical images/layers given identical
+  inputs, setting aside the problem of identifying those inputs.)
+
+## Target persona
+
+This feature will be useful for **buildpack authors** and **platform
+operators** towards tracing and reproducing images and layers.
+
+## Example
+
+Consider some buildpack `yabba` and some application `foo-api`. Let's
+also say that the `yabba` buildpack generates a cacheable build layer
+`yabba/build`.
+
+1. I do a build of the `foo-api` as of git SHA aaa000 involving the
+   `yabba` buildpack, without using any cache.
+2. I do a build of the `foo-api` as of git SHA aaa111 involving the
+   `yabba` buildpack, using the cache from the previous build.
+
+I want to account for all of the inputs that were present during the
+second build:
+
+1. the version of the `foo-api` application
+2. the version of the `yabba` buildpack
+3. the version of the stack's build image
+4. the version of the `yabba/build` layer that was restored from the
+   cache
+
+I can account for all but the last - I neither know that layer's
+identity, nor how it was created (the version of the `foo-api`, the
+version of the `yabba` buildpack, whatever layers existed at that
+time). I cannot identify the second image's inputs - in other words, I
+cannot trace the second image.
+
+Now consider that I have access to each image's `report.toml`, and
+that the second image's `report.toml` looks something like this:
+
+```toml
```

Comment:
I suggest removing the `toml` here because the conents aren't valid toml, and it makes the diff hard to read. But I can't figure out how to put just three backticks inside of a code block :)